samza-core/src/main/scala/org/apache/samza/config/StreamConfig.java

```diff
@@ -0,0 +1,394 @@
+    }
+
+    return streamIds.isEmpty() ? null : streamIds.get(0);
+  }
+
+  /**
+   * Gets the specified Samza property for a SystemStream. A Samza property is a property that controls how Samza
+   * interacts with the stream, as opposed to a property of the stream itself.
+   *
+   * First, tries to map the systemStream to a streamId. This will only find a streamId if the stream is a physical name
+   * (explicitly mapped physical name or a stream id without a physical name mapping). That means this will not map a
+   * stream id to itself if there is a mapping from the stream id to a physical stream name. This also requires that the
+   * stream id is mapped to a system in the config.
+   * If a stream id is found:
+   * 1) Look for "streams.{streamId}.{property}" for the stream id.
+   * 2) Otherwise, look for "systems.{systemName}.streams.{streamName}.{property}" in which the systemName is the system
+   * mapped to the stream id and the streamName is the physical stream name for the stream id.
+   * 3) Otherwise, look for "systems.{systemName}.default.stream.{property}" in which the systemName is the system
+   * mapped to the stream id.
+   * If a stream id was not found or no property could be found using the above keys:
+   * 1) Look for "systems.{systemName}.streams.{streamName}.{property}" in which the systemName is the system in the
+   * input systemStream and the streamName is the stream from the input systemStream.
+   * 2) Otherwise, look for "systems.{systemName}.default.stream.{property}" in which the systemName is the system
+   * in the input systemStream.
+   *
+   * @param systemStream the SystemStream for which the property value will be retrieved.
+   * @param property the samza property key (including the "samza." prefix); for example, for both
+   *                 "streams.streamId.samza.prop.key" and "systems.system.streams.streamName.samza.prop.key", this
+   *                 argument should have the value "samza.prop.key"
+   */
+  protected String getSamzaProperty(SystemStream systemStream, String property) {
+    if (!property.startsWith(SAMZA_PROPERTY)) {
+      throw new IllegalArgumentException(
+        String.format("Attempt to fetch a non samza property for SystemStream %s named %s", systemStream, property));
+    }
+
+    String streamVal = getAllStreamProperties(systemStreamToStreamId(systemStream)).get(property);
+    return (streamVal != null) ? streamVal : getSystemStreamProperties(systemStream.getSystem(), systemStream.getStream()).get(property);
+  }
+
+  /**
+   * Gets the System associated with the specified streamId.
+   * It first looks for the property
+   * streams.{streamId}.system
+   * <p>
+   * If no value was provided, it uses
+   * job.default.system
+   *
+   * @param streamId the identifier for the stream in the config.
+   * @return the system name associated with the stream.
```

Comment:
Could you update this javadoc to specify that it might also return null?
I actually think there are some callers of `getSystem` which don't really check for null cleanly, but probably not in scope to change that in this PR.